tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala

```diff
@@ -30,37 +41,56 @@ case class RandomForest(allOutcomes: Seq[Int], decisionTrees: Seq[DecisionTree])
     * @param featureVector feature vector to find outcome distribution for
     * @return a probability distribution over outcomes
     */
-  override def outcomeDistribution(featureVector: FeatureVector): Map[Int, Float] = {
-    val outcomeHistogram = decisionTrees map { decisionTree =>
-      decisionTree.classify(featureVector)
-    } groupBy { x => x } mapValues { v => v.size }
-    RandomForest.normalizeHistogram(outcomeHistogram)
-  }
+  override def outcomeDistribution(
+    featureVector: FeatureVector
+  ): (OutcomeDistribution, Option[Justification]) = {
 
-  /** An experimental weighted version of the above .outcomeDistribution method.
-    *
-    * @param featureVector feature vector to find outcome distribution for
-    * @return a probability distribution over outcomes
-    */
-  def outcomeDistributionAlternate(featureVector: FeatureVector): Map[Int, Float] = {
-    val summedOutcomeHistograms: Map[Int, Int] = decisionTrees flatMap { decisionTree =>
-      decisionTree.outcomeHistogram(featureVector).toSeq
-    } groupBy { case (x, y) => x } mapValues { case x => x map { _._2 } } mapValues { _.sum }
-    RandomForest.normalizeHistogram(summedOutcomeHistograms)
+    val decisionTreeOutputs: Seq[(Int, Option[Justification])] = decisionTrees map { decisionTree =>
+      decisionTree.classify(featureVector)
+    }
+    val outcomeHistogram = decisionTreeOutputs map {
+      _._1
+    } groupBy { x =>
+      x
+    } mapValues { v =>
+      v.size
+    }
+    val (bestOutcome, _) = outcomeHistogram maxBy { case (_, numVotes) => numVotes }
+    val majorityJustifications: Seq[(Int, Justification)] =
+      (decisionTreeOutputs.zipWithIndex filter {
+        case ((outcome, _), _) =>
+          outcome == bestOutcome
+      } map {
+        case ((_, maybeJustification), treeIndex) =>
+          maybeJustification map { justification =>
+            (treeIndex, justification)
+          }
+      }).flatten
+    val justification =
+      if (majorityJustifications.isEmpty) {
```

Comment:
Oh wait, no. This is empty if the underlying DecisionTree implementation you're using doesn't support justifications (i.e. it would be a flattened sequence of None, which is the empty sequence).